src/Catalyst.KBucket/KBucket.cs

```diff
@@ -0,0 +1,503 @@
+        ///   
+        ///   If the selection is our new contact, the old contact is removed and the new
+        ///   contact is marked as most recently contacted.
+        /// </remarks>
+        private void _Update(Bucket<T> node, int index, T contact)
+        {
+            var incumbent = node.Contacts[index];
+            var selection = Arbiter(incumbent, contact);
+
+            // if the selection is our old contact and the candidate is some new
+            // contact, then there is nothing to do
+            if (selection == incumbent && incumbent != contact) return;
+
+            node.Contacts.RemoveAt(index);
+            node.Contacts.Add(selection);
+        }
+
+        /// <summary>
+        ///   Determines whether the id at the bitIndex is 0 or 1.
+        /// </summary>
+        /// <returns>
+        ///   Left leaf if `id` at `bitIndex` is 0, right leaf otherwise
+        /// </returns>
+        /// <remarks>
+        ///   This is an internal method.  It should not be directly called.
+        ///   It is only public for unit testing.
+        /// </remarks>
+        public Bucket<T> _DetermineNode(Bucket<T> node, byte[] id, int bitIndex)
+        {
+            // id's that are too short are put in low bucket (1 byte = 8 bits)
+            // (bitIndex >> 3) finds how many bytes the bitIndex describes
+            // bitIndex % 8 checks if we have extra bits beyond byte multiples
+            // if number of bytes is <= no. of bytes described by bitIndex and there
+            // are extra bits to consider, this means id has less bits than what
+            // bitIndex describes, id therefore is too short, and will be put in low
+            // bucket
+            var bytesDescribedByBitIndex = bitIndex >> 3;
+            var bitIndexWithinByte = bitIndex % 8;
+            if (id.Length <= bytesDescribedByBitIndex && bitIndexWithinByte != 0) return node.Left;
+
+            // byteUnderConsideration is an integer from 0 to 255 represented by 8 bits
+            // where 255 is 11111111 and 0 is 00000000
+            // in order to find out whether the bit at bitIndexWithinByte is set
+            // we construct (1 << (7 - bitIndexWithinByte)) which will consist
+            // of all bits being 0, with only one bit set to 1
+            // for example, if bitIndexWithinByte is 3, we will construct 00010000 by
+            // (1 << (7 - 3)) -> (1 << 4) -> 16
+            var byteUnderConsideration = id[bytesDescribedByBitIndex];
+            return 0 != (byteUnderConsideration & (1 << (7 - bitIndexWithinByte))) ? node.Right : node.Left;
+        }
+
+        /// <summary>
+        ///   Get a contact by its exact ID. 
+        /// </summary>
+        /// <param name="id">
+        ///   The ID of a <see cref="IContact"/>.
+        /// </param>
+        /// <returns>
+        ///   <b>null</b> or the found contact.
+        /// </returns>
+        private T _Get(byte[] id)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Rename method '_Get' to match pascal case naming rules, trim underscores from the name.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)